### PR TITLE
Move mavlink reboot up to base class

### PR DESCRIPTION
**Pull request by peterbarker**
https://github.com/ArduPilot/ardupilot/pull/6640 solved a bug for Sub only; making the code common means everybody can benefit from this sort of fix.

However, the different vehicles currently have several things different between them:
 - Sub and Copter both poked AP_Notify that a reboot was occurring.  No harm in the other vehicles also getting that behaviour
- Plane pokes px4io to disable the mixers.  I've assumed that that is still only relevant for Plane and controlled it with a boolean "should" callback.
 - Sub had some interesting code which iterated through all of the outputs (badly; should be using _servo_count or similar in RCOutput, not NUM_RC_CHANNELS!)
   - I've retained this, also protected by a callback.

We need to work out what everybody should be doing here.  Should all vehicles be doing the same thing when rebooting in terms of mixing and Sub's "no short pwm signal" code?

If these should be the same on every vehicle, should we have a call on AP_HAL to do this "reboot safing"?  Possibly `AP_HAL::reboot` would call that?  At the moment a CLI reboot won't do any of these "safing" procedures, so a reboot in the middle of a motor-test might do interesting things...

### CODE REVIEW

**Comment by peterbarker** [comment]
Removing WIP.  This doesn't change function and I don't think it is going to act as a talking point for making sure all our vehicles do the same thing on reboot.

**Comment by khancyr** [comment]
LGTM, I thkin all vehicle should disable override and set rc to 0 on reboot, just to prevent error , like on rover with brushed pwm ... 1500us isn't 0 ...

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -1628,7 +1628,8 @@ uint8_t GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_t &packe
             }
 #endif
         }
-
+        // send ack before we reboot
+        mavlink_msg_command_ack_send(chan, packet.command, MAV_RESULT_ACCEPTED);
```
Comment (lucasdemarchi) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: do we have any guarantee here the buffer will be flushed?

**Comment by lucasdemarchi** [comment]
small comment above, but looks good.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -1608,10 +1620,10 @@ void GCS_MAVLINK::send_accelcal_vehicle_position(uint32_t position)
   motors. That can be dangerous when a preflight reboot is done with
   the pilot close to the aircraft and can also damage the aircraft
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: Come to think of it - what about the NavIO2 board?  It has an IO processor, but this code is guarded for vrbrain/px4

Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: Ah... but reboot() probably isn't supported for it.

Comment (lucasdemarchi) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: we should implement it though... probably a reexec. The only thing that would be compatible would probably be the ESC cal

**Comment by peterbarker** [comment]
On Thu, 3 Aug 2017, Lucas De Marchi wrote:

> we should implement it though... probably a reexec. The only thing that
> would be compatible would probably be the ESC cal

Yes, I actually looked at this for SITL :-)

The first problem I hit was this:

bind failed on port 5760 - Address already in use

That's not going to be a problem for Linux, but the other bound ports will 
be :-)


https://github.com/peterbarker/ardupilot/tree/sitl-reboot-support

(also obviously missing close-extra-fds where "extra" needs to be defined)

**Comment by lucasdemarchi** [comment]
> 
bind failed on port 5760 - Address already in use

That's just a flag we need to pass to the socket

> 
(also obviously missing close-extra-fds where "extra" needs to be defined)

I did a pass on the code recently adding the O_CLOEXEC flag. Now there may be a few more ... But it's also easy to detect by looking at open fds after exec

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 2 above)
Comment (OXINARF) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: @peterbarker The guarded code is different I think. It will stop the IO co-processor from taking over on what is output. AFAIK, the IO processor in NavIO2 board does the IO processing, but doesn't take decisions of its own, like the one in PX4 boards - what it needs is for us to stop outputting PWM. I'm not sure why writing 0 to the RCOutput is considered a _don't update_ feature, but the workaround to use 1 (that Sub implemented) should be enough.

**Comment by peterbarker** [comment]
Outcomes from the call:
 - all vehicles should disable overrides when rebooting
  - Plane should NOT disable overrides if the vehicle is armed
  - if other vehicles get px4io mixer firmware which does sensible things then they could also not disable overrides
 - for various reasons, sending a value of 1 to the output channels on reboot is not a good idea, and what Sub is doing shouldn't spread

The former is a relatively striaght-forward thing.  Fixing the latter is a bit of digging.

**Comment by peterbarker** [comment]
@OXINARF I've pushed up a change so that all vehicles will disable overrides by default, EXCEPT plane which (unless it is a quadplane) will allow overrides to happen.

I'm uneasy with changing the behaviour of Plane to disable overrides when not armed.  Notionally if you were to disarm in flight at the moment and *then* reboot you could still glide the vehicle.  I've no idea why you would do that.....

Is this mergeable now?

**Comment by OXINARF** [comment]
@peterbarker You are welcome to bring this to a dev call and see if people want to merge it (or even merge it by yourself, you have the permissions to merge stuff). Personally, I don't like it. I feel that we introduce several methods just to keep differences between vehicles and so, in that case, I don't see value in this change - it makes harder knowing what each vehicle does on a reboot.

I still think that all vehicles should do the same on reboot - with the caveat that in my opinion we shouldn't allow a reboot once a vehicle is armed.
Regarding your example of unarmed and still flying, I think it wouldn't work if we did what was discussed in the dev call, also reboot the IO processor - and that's what a real reboot would be.

In conclusion, I'm not going to oppose it going in if people want it, but, as is, I don't see value in it so I won't merge it without being told to.

**Comment by peterbarker** [comment]
On Thu, 14 Sep 2017, Francisco Ferreira wrote:

> I still think that all vehicles should do the same on reboot - with the

I think Sub's special handling code isn't great, and should go away.  But 
Plane's special handling is different - special effort has been made so 
a reboot of the FMU in-flight doesn't mean vehicle death.

> caveat that in my opinion we shouldn't allow a reboot once a vehicle is
> armed.

Well, the command *is* "preflight reboot", so that would be reasonable.

I'm not trying to change behaviour, just trying to share code so we can 
have the vehicles be more similar in their behaviours in future.  Notably 
working out exactly why sub needs to do what it is doing and share that 
with all vehicles if required (or remove it if it is not required!)

**Comment by peterbarker** [comment]
@jaxxzer Where does the "ignore 0-values on servo outputs" hack lie in the source code?

**Comment by jaxxzer** [comment]
It's been a long time since I looked at this.
I \*think\* the ignore-zero bit is here https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_PX4/RCOutput.cpp#L530

**Comment by jaxxzer** [comment]
If it isn't clear the problem is that Sub uses reversible ESCs, so pulse lengths of 1500us are neutral, and pulse lengths between 1100us and 1500us are commands to spin. When the vehicle was armed, and I was flashing new code with px_uploader.py, occasionally the motors would spin up very briefly as the vehicle rebooted into the bootloader. This appeared that sometimes, the 1500us pulse got cut short somewhere between 1100 and 1500 causing the motors to spin. The resolution was to send pulses lower than 1100 (invalid) before rebooting.

**Comment by jaxxzer** [comment]
I'm actually not entirely sure if this was actually the problem, and I don't remember if the problem was solved this way (I think it was). I do not do this nearly as often anymore now that I've got SITL :laughing:.

**Comment by peterbarker** [comment]
On Fri, 15 Sep 2017, jaxxzer wrote:

> I'm actually not entirely sure if this was actually the problem, and I don't
> remember if the problem was solved this way (I think it was). I do not do
> this nearly as often anymore now that I've got SITL 😆.

I haven't found were "0" is special-cased.  FF found the same place I did 
where max-uint16_t was used...

We'd like to use this zero-outputs stuff, but tridge pointed out on the 
call that an output value of 1 can do Bad Things.

**Comment by jaxxzer** [comment]
I think the motivation for using 1 instead of zero might have been either something I saw in the px4 layer ( I get really lost in there ), or just the comment here: https://github.com/ArduPilot/PX4Firmware/blob/0384802e81330751f39550a0804357d8b04f278a/src/drivers/drv_pwm_output.h#L40

In any case, I do recall trying zero first, and it did not work, the motors still spun, and I changed it to one afterwards. It's been more than a year...

**Comment by amilcarlucas** [comment]
This probably needs a rebase

**Comment by amilcarlucas** [comment]
@peterbarker this needs a rebase

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 3]
```diff
@@ -1630,10 +1642,10 @@ void GCS_MAVLINK::send_accelcal_vehicle_position(uint32_t position)
...
  */
-MAV_RESULT GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_t &packet, bool disable_overrides)
+MAV_RESULT GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_t &packet)
 {
     if (is_equal(packet.param1,1.0f) || is_equal(packet.param1,3.0f)) {
```
Comment (lucasdemarchi) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: it would be much nicer if this if was inverted and we returned early. While at it, fix the coding style to have spaces

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 4]
```diff
@@ -1650,7 +1662,17 @@ MAV_RESULT GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_t &pa
             }
 #endif
         }
-
+        if (should_zero_rc_outputs_on_reboot()) {
+            reboot_zero_rc_outputs();
```
Comment (lucasdemarchi) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: humn... the method should be `zero_rc_outputs()` or the `should_zero_rc_outputs_on_reboot()` should be moved inside the method. Otherwise it is redundant

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 5]
```diff
@@ -1650,7 +1662,17 @@ MAV_RESULT GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_t &pa
...
+        }
+        // send ack before we reboot
+        mavlink_msg_command_ack_send(chan, packet.command, MAV_RESULT_ACCEPTED);
```
Comment (lucasdemarchi) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: we should make sure to flush the tx buffers. Relying on a delay below to do that is not good IMO.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 6]
```diff
@@ -1622,6 +1622,18 @@ void GCS_MAVLINK::send_accelcal_vehicle_position(uint32_t position)
...
+void GCS_MAVLINK::reboot_zero_rc_outputs()
+{
+    // Send an invalid signal to the motors to prevent spinning due to neutral (1500) pwm pulse being cut short
+    // For that matter, send an invalid signal to all channels to prevent undesired/unexpected behavior
+    hal.rcout->cork();
+    for (int i=0; i<NUM_RC_CHANNELS; i++) {
+        // Set to 1 because 0 is interpreted as flag to ignore update
+        hal.rcout->write(i, 1);
+    }
+    hal.rcout->push();
```
Comment (lucasdemarchi) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: SRV_Channels::push() and RCOutput::push() are not the same thing.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 5 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Is there a good way to do that from the main thread, 'though?  This is a no-worse-off-and-possibly-better patch as-is.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Renamed it.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Pushing a patch on top which tidies it all up.

**Comment by peterbarker** [comment]
@jaxxzer I'd like to set the rc outputs to `0` rather than `1`.  To that end I've pushed a patch on top which sends 0 rather than 1.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Fixed.

**Comment by peterbarker** [comment]
@jaxxzer This can go in once it is tested - can you do that, please?

**Comment by peterbarker** [comment]
Two analyzer traces.  The first is what happens when we do not zero RC outputs on reboot.  The second is whehn we do zero rc outputs on reboot.  I copied Sub's should_zero_rc_outputs_on_reboot into Copter between these two traces.

This was on a PixHawk.

Saelee Channel 0 is main output 5 on the PixHawk
Saelee Channel 1 is main output 1 on the PixHawk

Vehicle was armed and safety off, with "props spinning" at MOT_SPIN_ARMED - that's why channel 1 has output.

servo channel 5 was configured as RC pass-through, with MAVProxy specifying 1800pwm.

Before these changes main channel 5 continues to pass through the RC input value - about another 500ms longer than main channel 1 does.


before-jacob

after-jacob

**Comment by peterbarker** [comment]
Oh.  Saelee channel 2 is aux-out 1...

I think the traces show that setting zero into those outputs is having an effect.  Losing the 500ms difference between main-output-1 and main-output-5 is a side-effect.

**Comment by tridge** [comment]
pending testing by @jaxxzer

**Comment by jaxxzer** [comment]
Everything appears ok here except for set_servo output extends after the reboot command.

The original idea was to ensure the output would not be stopped mid-pulse. We need to finish a complete pulse at the required width, then stop sending pulses. I looked between master and this pr, and could not see any pulses being cut short in any case.

**Comment by jaxxzer** [comment]
I think this is ok to go in.
